docs/GettingStartedDocs/best_practices_secure_enclaves.md

```diff
@@ -0,0 +1,380 @@
+// pitfalls.
+int ecall_with_user_check0(void* ptr) {
+
+    blob_t* blob;
+
+    if (ptr == nullptr)
+        return -1;
+
+    blob = (blob_t*)ptr;
+
+    // Blobs to be rendered are limited to 4K.
+    if (blob->size > 4096)
+        return -1;
+
+    return render_data(blob->data, blob->size);
+}
+```
+
+## 1.1 Ensuring memory is where it should be
+
+Remember that the code we are focused on is running _inside_ of the enclave.
+The caller invoked this ecall function from _outside_ the enclave, passing in a
+pointer to memory containing well-formed data (in the non-malicious case).
+But in the Open Enclave security model, the ecall invoker is untrusted - we must
+treat untrusted input with caution.
+
+One important security check is to ensure that memory buffers are on the correct
+side of the security boundary. The Open Enclave SDK provides two library
+functions for checking memory buffers:
+`oe_is_within_enclave()` and `oe_is_outside_enclave()`.
+
+Let's update the code to ensure that the memory buffer is at least the size of a
+`blob_t` and is located _strictly_ outside the secured enclave memory region
+before operating on it by calling `oe_is_outside_enclave()`.
+
+```diff
+// WARNING: Portions of this code are intentionally flawed to demonstrate common
+// pitfalls.
+int ecall_with_user_check1(void* ptr) {
+
+    blob_t* blob;
+
+-   if (ptr == nullptr)
+-       return -1;
++   // Ensure passed-in pointer is not null and the memory buffer is located
++   // entirely outside of the secured enclave memory region.
++   if (!oe_is_outside_enclave(ptr, sizeof(blob_t)))
++       return -1;
+
+    blob = (blob_t*)ptr;
+
+    // Blobs to be rendered are limited to 4K.
+    if (blob->size > 4096)
+        return -1;
+
+    return render_data(blob->data, blob->size);
+}
+```
+
+## 1.2 Memory bounds: oe_is_within_enclave vs. oe_is_outside_enclave
+
+Let's highlight a logical mistake that we've seen made in enclave applications.
+Some have regarded the memory validation functions as boolean opposites of each
+other, that is: `oe_is_within_enclave() == !oe_is_outside_enclave()`. This is
+incorrect for several reasons.
+
+To understand why these functions are _not_ boolean opposites, let's examine the
+Open Enclave implementation for SGX as an example:
+[memory.c](https://github.com/openenclave/openenclave/blob/08ebe60c1d2e3ea24ac634c673a420296bff3352/enclave/core/sgx/memory.c).
+There are three conditions that the functions validate:
+1)	The pointer is not null.
+2)	The bounding arithmetic operations do not wrap (that is, numerical
+overflow).
+3)	The buffer lies completely within or outside of the secured enclave memory
+region.
+
+The first two conditions should make clear the pitfalls of considering the
+function pair to be boolean opposites: If the pointer is null or the bounding
+calculations overflow in a call to `oe_is_within_enclave()`, the function will
+return `false`, which should _not_ be interpreted as "the buffer is outside the
```

Comment:
```suggestion
return `false`, which should _not_ be interpreted as "the buffer is strictly outside the
```
for consistency with line 114.  Or use "entirely" for consistency with line 229.  For bonus points, make 114 and 229 use the same word.

Reply:
Again, good catch.  I adopted "strictly" throughout the doc to align with the Open Enclave API documentation.

Reply:
I also reduced "memory region" to "memory".